integration/builder.go

```diff
@@ -0,0 +1,121 @@
+package integration
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"sync"
+
+	"github.com/newrelic/infra-integrations-sdk/args"
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/infra-integrations-sdk/persist"
+	"github.com/pkg/errors"
+)
+
+const protocolVersion = "2"
+
+// Builder OOP builder-pattern to create a new Integration instance.
+type Builder struct {
```

Comment:
Rather than implementing a builder pattern here, I would vote for using functional options. Well explained in Dave Chenney's blog https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis.

`func New(name, version string, opts ...func(*Integration)`

Reply:
Of course we could create a bunch of predefined functions for setting the Writer, etc.

Reply:
I like it, but I don't think it's as understandable as the builder pattern.

Reply:
The builder pattern is understandable if you're familiar with it. From the point of view of a developer of an integration, I would prefer using something like:
`integration.New("redis", "2.0", integration.HTTPWriter)`
rather than:
`integration.NewBuilder("redis", "2.0").Writer(integration.HTTPWriter).Build()`

In addition to that, personally I feel `NewBuilder` confusing (we already have the concept of nr-integration-builder). 

Also, we should think that usually people will use the options that we provide, I'm not sure if they are going to write a lot of functional options by themselves. Otherwise, I don't consider it's hard to understand functional options approach.

Reply:
We allow the user to replace these services via fluent api:

* Logger
* ParsedArguments (not service but args)
* Synchronized (not service, just enables concurrency)
* Writer
* Storer
* InMemoryStore (replaces Storer by builtin in-memory)

You would need lots of different constructors or options to enable this flexibility. So I'd understand either:
* use a builder
* use a factory
* use functional options
* remove flexibility (services/options)

But I don't see how to easily do this adding several constructor, neither using config options (as this won't allow to customize services)